src/cluster.c

```diff
@@ -2823,64 +2792,76 @@ valkeyClusterGetValkeyAsyncContext(valkeyClusterAsyncContext *acc,
+    valkeyClusterAsyncSetError(acc, cc->err, cc->errstr);
+
+    if (options->async_connect_cb != NULL) {
+        acc->onConnect = options->async_connect_cb;
+    }
+    if (options->async_disconnect_cb != NULL) {
+        acc->onDisconnect = options->async_disconnect_cb;
+    }
+    if (options->attach_fn != NULL) {
+        acc->attach_fn = options->attach_fn;
+        acc->attach_data = options->attach_data;
+    }
 
     return acc;
 }
 
-valkeyClusterAsyncContext *valkeyClusterAsyncConnect(const char *addrs,
-                                                     int flags) {
-
-    valkeyClusterContext *cc;
-    valkeyClusterAsyncContext *acc;
-
-    cc = valkeyClusterConnect(addrs, flags);
-    if (cc == NULL) {
-        return NULL;
-    }
-
-    acc = valkeyClusterAsyncInitialize(cc);
+valkeyClusterAsyncContext *valkeyClusterAsyncConnectWithOptions(const valkeyClusterOptions *options) {
+    valkeyClusterAsyncContext *acc = valkeyClusterAsyncContextInit(options);
     if (acc == NULL) {
-        valkeyClusterFree(cc);
         return NULL;
     }
 
+    valkeyClusterAsyncConnect(acc);
     return acc;
 }
 
-int valkeyClusterAsyncConnect2(valkeyClusterAsyncContext *acc) {
+int valkeyClusterAsyncConnect(valkeyClusterAsyncContext *acc) {
     /* An attach function for an async event library is required. */
     if (acc->attach_fn == NULL) {
         return VALKEY_ERR;
     }
-    return updateSlotMapAsync(acc, NULL /*any node*/);
-}
 
-int valkeyClusterAsyncSetConnectCallback(valkeyClusterAsyncContext *acc,
-                                         valkeyConnectCallback *fn) {
-    if (acc->onConnect != NULL)
-        return VALKEY_ERR;
-    acc->onConnect = fn;
-    return VALKEY_OK;
+    /* Clear a previously set shutdown flag to allow a
+     * reconnection of an async context using this API. */
+    acc->cc->flags &= ~VALKEY_FLAG_DISCONNECTING;
+
+    /* Use blocking initial slotmap update when configured. */
+    if (acc->cc->flags & VALKEY_FLAG_BLOCKING_INITIAL_UPDATE) {
+        if (valkeyClusterUpdateSlotmap(acc->cc) != VALKEY_OK) {
+            valkeyClusterAsyncSetError(acc, acc->cc->err, acc->cc->errstr);
+            return VALKEY_ERR;
+        }
+        return VALKEY_OK;
+    }
+    /* Use non-blocking initial slotmap update. */
+    return updateSlotMapAsync(acc, NULL /*any node*/);
 }
 
-int valkeyClusterAsyncSetDisconnectCallback(valkeyClusterAsyncContext *acc,
-                                            valkeyDisconnectCallback *fn) {
-    if (acc->onDisconnect == NULL) {
-        acc->onDisconnect = fn;
+int valkeyClusterAsyncSetEventCallback(valkeyClusterAsyncContext *acc,
+                                       void(fn)(const valkeyClusterContext *cc,
+                                                int event, void *privdata),
+                                       void *privdata) {
+    if (acc->cc->event_callback == NULL) {
+        acc->cc->event_callback = fn;
+        acc->cc->event_privdata = privdata;
```

Comment:
Can be done using the options object? Or not?

Reply:
`valkeyClusterAsyncContextInit` and `valkeyClusterAsyncSetEventCallback` are kept since we need to be able to use the created async-context when setting `privdata` in the event callback., i.e. chicken or the egg problem..
Example use-cases are when waiting for an VALKEYCLUSTER_EVENT_READY before sending the first async command in the event callback, using the async-context received in privdata.

I would be nice to get rid of them, but I have not yet found a solution for it.

Reply:
OK, let's keep it.